src/simulators/state_chunk.hpp

```diff
@@ -0,0 +1,976 @@
+            continue;
+          }
+          else{
+            iLocalChunk = iChunk1;
+            iRemoteChunk = iChunk2;
+            iProc = get_process_by_chunk(iChunk2);
+          }
+        }
+        else{
+          if(iChunk2 >= chunk_index_begin_[distributed_rank_] && iChunk2 < chunk_index_end_[distributed_rank_]){    //chunk2 is on this process
+            iLocalChunk = iChunk2;
+            iRemoteChunk = iChunk1;
+            iProc = get_process_by_chunk(iChunk1);
+          }
+          else{
+            continue;   //there is no chunk for this pair on this process
+          }
+        }
+
+        MPI_Request reqSend,reqRecv;
+        MPI_Status st;
+        uint_t sizeRecv,sizeSend;
+
+        void* pSend = qregs_[iLocalChunk - global_chunk_index_].send_buffer(sizeSend);
+        MPI_Isend(pSend,sizeSend,MPI_BYTE,iProc,iPair,distributed_comm_,&reqSend);
+
+        void* pRecv = qregs_[iLocalChunk - global_chunk_index_].recv_buffer(sizeRecv);
+        MPI_Irecv(pRecv,sizeRecv,MPI_BYTE,iProc,iPair,distributed_comm_,&reqRecv);
+
+        MPI_Wait(&reqSend,&st);
+        MPI_Wait(&reqRecv,&st);
+        //MPI_Sendrecv can be used if number of processes = 2^m
+        //MPI_Sendrecv(pSend,sizeSend,MPI_BYTE,iProc,distributed_rank_,pRecv,sizeRecv,MPI_BYTE,iProc,iProc,MPI_COMM_WORLD,&st);
+
+        qregs_[iLocalChunk - global_chunk_index_].apply_chunk_swap(qubits,iRemoteChunk);
+      }
+    }
+#endif
+
+  }
+}
+
+template <class state_t>
+void StateChunk<state_t>::reduce_sum(rvector_t& sum) const
+{
+#ifdef AER_MPI
+  if(distributed_procs_ > 1){
+    uint_t i,n = sum.size();
+    rvector_t tmp(n);
+    MPI_Allreduce(&sum[0],&tmp[0],n,MPI_DOUBLE_PRECISION,MPI_SUM,distributed_comm_);
+    for(i=0;i<n;i++){
+      sum[i] = tmp[i];
+    }
+  }
+#endif
+}
+
+template <class state_t>
+void StateChunk<state_t>::reduce_sum(complex_t& sum) const
+{
+#ifdef AER_MPI
+  if(distributed_procs_ > 1){
+    complex_t tmp;
+    MPI_Allreduce(&sum,&tmp,2,MPI_DOUBLE_PRECISION,MPI_SUM,distributed_comm_);
+    sum = tmp;
+  }
+#endif
+}
+
+template <class state_t>
+void StateChunk<state_t>::reduce_sum(double& sum) const
+{
+#ifdef AER_MPI
+  if(distributed_procs_ > 1){
+    double tmp;
+    MPI_Allreduce(&sum,&tmp,1,MPI_DOUBLE_PRECISION,MPI_SUM,distributed_comm_);
+    sum = tmp;
+  }
+#endif
+}
+
+template <class state_t>
+void StateChunk<state_t>::gather_value(rvector_t& val) const
+{
+#ifdef AER_MPI
+  if(distributed_procs_ > 1){
+    MPI_Alltoall(&val[0],1,MPI_DOUBLE_PRECISION,&val[0],1,MPI_DOUBLE_PRECISION,distributed_comm_);
+  }
+#endif
+}
+
+template <class state_t>
+void StateChunk<state_t>::sync_process(void) const
+{
+#ifdef AER_MPI
+  if(distributed_procs_ > 1){
+    MPI_Barrier(distributed_comm_);
+  }
+#endif
+}
+
+//gather distributed state into vector (if memory is enough)
+template <class state_t>
+void StateChunk<state_t>::gather_state(std::vector<std::complex<double>>& state)
```

Comment:
This is because State classes can not specify `data_t` so the vector is initialized as `auto` at the caller side (see `add_state_to_data` function in statevector_state_chunk.hpp)